src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/GoogleDataprocDAO.scala

```diff
@@ -1,79 +1,241 @@
+    try {
+      executeGoogleRequest(request) // returns a Firewall
+    } catch {
+      case e: GoogleJsonResponseException =>
+        if (e.getStatusCode == 404)
+          addFirewallRule(googleProject)
+        else
+          throw new FirewallRuleInaccessibleException(googleProject, dataprocConfig.clusterFirewallRuleName)
+    }
   }
 
-  private def build(googleProject: String, clusterName: String, clusterRequest: ClusterRequest)(implicit executionContext: ExecutionContext): Future[Operation] = {
+  /* Adds a firewall rule in the given google project. This firewall rule allows ingress traffic through a specified port for all
+     VMs with the network tag "leonardo". This rule should only be added once per project.
+    To think about: do we want to remove this rule if a google project no longer has any clusters? */
+  private def addFirewallRule(googleProject: String): Future[ComputeOperation] = {
     Future {
-      //currently, the bucketPath of the clusterRequest are not used - it will be used later as a place to store notebooks and results
-      val metadata = clusterRequest.labels + ("docker-image" -> dataprocConfig.dataprocDockerImage)
-      val gce = new GceClusterConfig().setMetadata(metadata.asJava).setServiceAccount(clusterRequest.serviceAccount)
-      val initActions = Seq(new NodeInitializationAction().setExecutableFile(dataprocConfig.dataprocInitScriptURI))
-      val clusterConfig = new ClusterConfig().setGceClusterConfig(gce).setInitializationActions(initActions.asJava)
-      val cluster = new Cluster().setClusterName(clusterName).setConfig(clusterConfig)
-      val request = dataproc.projects().regions().clusters().create(googleProject, dataprocConfig.dataprocDefaultZone, cluster)
+      // Create an Allowed object that specifies the port and protocol of rule
+      val allowed = new Allowed().setIPProtocol(proxyConfig.jupyterProtocol).setPorts(List(proxyConfig.jupyterPort.toString).asJava)
+      // Create a firewall rule that takes an Allowed object, a name, and the network tag a cluster must have to be accessed through the rule
+      val firewallRule = new Firewall()
+        .setName(dataprocConfig.clusterFirewallRuleName)
+        .setTargetTags(List(dataprocConfig.clusterNetworkTag).asJava)
+        .setAllowed(List(allowed).asJava)
+
+      val request = new Compute(httpTransport, jsonFactory, getServiceAccountCredential(vmScopes)).firewalls().insert(googleProject, firewallRule)
+
       try {
-        executeGoogleRequest(request)
+        executeGoogleRequest(request) // returns a ComputeOperation
+      } catch {
+        case e: GoogleJsonResponseException => throw new FirewallRuleInsertionException(googleProject, dataprocConfig.clusterFirewallRuleName)
+      }
+    }
+  }
+
+  /* Create init bucket for the cluster and populate it with the cluster initialization files */
+  private def initializeBucket(googleProject: GoogleProject, clusterName: String, bucketName: String) = {
```

Comment:
Currently, this doesn't have a return type. I would ideally like it to be `Future[(Bucket, Array[StorageObject])]`, but in the function, I have a `Future[Bucket]` and a `Future[Array[StorageObject]]`. I'm not exactly sure how wrangle these to things into a Future of a tuple.

Reply:
I was wrong, for-comprehensions do go in sequence, so I changed made it a for-comp